File: solutions/studyplan/programmingskills/easy_13/Solution.java
package studyplan.programmingskills.easy_13;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Objects;

public class Solution {

    /*  My Solution
     *  Runtime: 3ms
     *  Memory: 46.21MB
     */
    int eachNumber(char ch) {
        return switch (ch) {
            case 'I' -> 1;
            case 'V' -> 5;
            case 'X' -> 10;
            case 'L' -> 50;
            case 'C' -> 100;
            case 'D' -> 500;
            case 'M' -> 1000;
            default -> 0;
        };
    }

    int romanToInt(String roman) {
        int result = 0;
        for (int i = 0; i < roman.length(); i++) {
            if (i + 1 < roman.length() && eachNumber(roman.charAt(i + 1)) > eachNumber(roman.charAt(i))) {
                result -= eachNumber(roman.charAt(i));
            } else {
                result += eachNumber(roman.charAt(i));

            }
        }
        return result;
    }
    /*  GPT Solution
     *  Runtime: 3ms
     *  Memory: 46.38MB
     */
    int romanToInt2(String s) {
        int[] val = new int[256];
        val['I'] = 1;
        val['V'] = 5;
        val['X'] = 10;
        val['L'] = 50;
        val['C'] = 100;
        val['D'] = 500;
        val['M'] = 1000;

        int result = 0;

        for (int i = 0; i < s.length(); i++) {
            if (i + 1 < s.length() && val[s.charAt(i)] < val[s.charAt(i + 1)])
                result -= val[s.charAt(i)];
            else
                result += val[s.charAt(i)];
        }

        return result;
    }
}

File: solutions/studyplan/programmingskills/easy_13/Solution.kt
package studyplan.programmingskills.easy_13

import java.sql.Array

/*  My Solution
   *  Runtime: 5ms
   *  Memory: 46.10MB
   */
fun eachNumber(ch: Char): Int {
    return when (ch) {
        'I' -> 1
        'V' -> 5
        'X' -> 10
        'L' -> 50
        'C' -> 100
        'D' -> 500
        'M' -> 1000
        else -> 0
    }
}

fun romanToInt(s: String): Int {
    var result = 0;
    for (i in 0 until s.length  ) {
        if (i + 1 < s.length && eachNumber(s[i]) < eachNumber(s[i+1])) {
            result -= eachNumber(s[i]);
        } else {
            result += eachNumber(s[i]);
        }
    }
    return result;
}
/*  GPT Solution
   *  Runtime: 4ms
   *  Memory: 46.18MB
   */
fun romanToInt1(s: String): Int {
    val valMap = IntArray(256)
    valMap['I'.code] = 1
    valMap['V'.code] = 5
    valMap['X'.code] = 10
    valMap['L'.code] = 50
    valMap['C'.code] = 100
    valMap['D'.code] = 500
    valMap['M'.code] = 1000

    var result = 0

    for (i in s.indices) {
        if (i + 1 < s.length && valMap[s[i].code] < valMap[s[i + 1].code])
            result -= valMap[s[i].code]
        else
            result += valMap[s[i].code]
    }

    return result
}